xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import io.envoyproxy.envoy.config.rbac.v3.Permission;
+import io.envoyproxy.envoy.config.rbac.v3.Policy;
+import io.envoyproxy.envoy.config.rbac.v3.Principal;
+import io.envoyproxy.envoy.config.rbac.v3.RBAC;
+import io.envoyproxy.envoy.config.rbac.v3.RBAC.Action;
+import io.grpc.xds.AuthorizationEngine.AuthDecision.DecisionType;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Implementation of gRPC server access control based on envoy RBAC protocol:
+ * https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto
+ *
+ * <p>One GrpcAuthorizationEngine is initialized with one action type and a list of policies.
+ * Policies are examined sequentially in order in an any match fashion, and the first matched policy
+ * will be returned. If not matched at all, the opposite action type is returned as a result.
+ * The the following example demonstrates the matching criteria corresponding to the engine:
+ * action: ALLOW
+ * policies:
+ *   "service-admin":
+ *     permissions:
+ *       - any: true
+ *     principals:
+ *       - authenticated:
+ *           principal_name:
+ *             exact: "cluster.local/ns/default/sa/admin"
+ *       - authenticated:
+ *           principal_name:
+ *             exact: "cluster.local/ns/default/sa/superuser"
+ *   "product-viewer":
+ *     permissions:
+ *         - and_rules:
+ *             rules:
+ *               - header: { name: ":method", exact_match: "GET" }
+ *               - url_path:
+ *                   path: { prefix: "/products" }
+ *               - or_rules:
+ *                   rules:
+ *                     - destination_port: 80
+ *                     - destination_port: 443
+ *     principals:
+ *       - any: true
+ *
+ * <p>The matching criteria should be immutable once constructed, and
+ * {@link #evaluate(EvaluateArgs)} is supposed to be called safely for multiple times.
+ */
+public final class GrpcAuthorizationEngine implements AuthorizationEngine {
+  private final Map<String, PolicyMatcher> policyMatchers = new HashMap<>();
+  private final Action action;
+
+  /** Instantiated with envoy RBAC policy configuration. */
+  public GrpcAuthorizationEngine(RBAC rbacPolicy) {
```

Comment:
We don't want this class to depend on RBAC (since we are going through the trouble of having the intermediate encoding). Have the caller pass in `Map<String, PolicyMatcher>` (and not `Map<String, Matcher`).